deploy/report/api/main.py

```diff
@@ -0,0 +1,307 @@
+        cls, v: Dict[str, List[Any]], values: Dict[str, List[Any]], **kwargs: Any
+    ) -> Dict[str, List[Any]]:
+        """Check if metadata columns have the same length as preds_prob and target.
+
+        Parameters
+        ----------
+        v : Dict[str, List[Any]]
+            Dictionary of values.
+        values : Dict[str, List[Any]]
+            Dictionary of values.
+
+        Returns
+        -------
+        Dict[str, List[Any]]
+            Dictionary of values.
+
+        Raises
+        ------
+        ValueError
+            If metadata columns have different lengths than preds_prob and target.
+
+        """
+        if "preds_prob" in values:
+            for column in v.values():
+                if len(column) != len(values["preds_prob"]):
+                    raise ValueError(
+                        "All metadata columns must have the same length as preds_prob and target"
+                    )
+        return v
+
+
+# This endpoint serves the UI
+@app.get("/", response_class=HTMLResponse)
+async def get_home() -> HTMLResponse:
+    """Return home page for cyclops model report app.
+
+    Returns
+    -------
+    HTMLResponse
+        Home page for cyclops model report app.
+
+    """
+    return templates.TemplateResponse(
+        "test_report.html", {"request": {"method": "POST"}}
+    )
+
+
+@app.post("/evaluate")
+async def evaluate_result(data: EvaluationInput) -> None:
+    """Calculate metric and return result from request body.
+
+    Parameters
+    ----------
+    data : EvaluationInput
+        Input data for evaluation.
+
+    Raises
+    ------
+    HTTPException
+        If there is an internal server error.
+
+    """
+    try:
+        # Create a dictionary with all data
+        df_dict = {
+            "target": data.target,
+            "preds_prob": data.preds_prob,
+            **data.metadata,
+        }
+
+        # Create DataFrame
+        df = pd.DataFrame(df_dict)
+
+        _eval(df)
+        LOGGER.info("Generated report.")
+    except Exception as e:
+        LOGGER.error(f"Error during evaluation: {str(e)}")
+        raise HTTPException(status_code=500, detail="Internal server error") from e
+
+
+@app.get("/evaluate")
+async def get_report() -> HTMLResponse:
+    """Return latest updated model report.
+
+    Returns
+    -------
+    HTMLResponse
+        Latest updated model report as HTML response.
+
+    """
+    return templates.TemplateResponse(
+        "test_report.html", {"request": {"method": "GET"}}
+    )
+
+
+def _export(report: ModelCardReport) -> None:
+    """Prepare and export report file."""
+    if not os.path.exists("./cyclops_report"):
+        LOGGER.info("Creating report for the first time!")
+    report_path = report.export(
+        output_filename="test_report.html",
+        synthetic_timestamp=str(datetime.today()),
+        last_n_evals=3,
+    )
+    shutil.copy(f"{report_path}", TEMPLATES_PATH)
+
+
+def _eval(df: pd.DataFrame) -> None:
```

Comment:
Is there a way to generalize this to a user-definable callback?

Reply:
Will try to do that in a separate PR later